pyproject.toml

```diff
@@ -199,13 +199,14 @@ authlib = { version = ">=0.14.0", optional = true }
 systemd-python = { version = ">=231", optional = true }
 lxml = { version = ">=4.2.0", optional = true }
 sentry-sdk = { version = ">=0.7.2", optional = true }
-opentracing = { version = ">=2.2.0", optional = true }
-jaeger-client = { version = ">=4.0.0", optional = true }
 txredisapi = { version = ">=1.4.7", optional = true }
 hiredis = { version = "*", optional = true }
 Pympler = { version = "*", optional = true }
 parameterized = { version = ">=0.7.4", optional = true }
 idna = { version = ">=2.5", optional = true }
+opentelemetry-api = {version = "^1.12.0", optional = true}
+opentelemetry-sdk = {version = "^1.12.0", optional = true}
+opentelemetry-exporter-jaeger = {version = "^1.12.0", optional = true}
```

Comment:
Asked from help in #synapse-dev, https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$uKejWRmZJyURmueN-uMegnBC2sTiOaru_iEh7jDh4UU?via=matrix.org&via=element.io&via=beeper.com

> I have no idea what I am doing wrong with trying to add some new dependencies. Trying to get the [OpenTelemetry PR](https://github.com/matrix-org/synapse/pull/13400) in shape again and get a fresh poetry lock reconciled with `opentelemetry-api`, `opentelemetry-sdk`, `opentelemetry-exporter-jaeger` but I never end up with the dependencies.
> 
> My end goal is:
> 
>  1. Remove `jaeger-client`, `opentracing` and `types-opentracing`
>  1. Install `opentelemetry-api`, `opentelemetry-sdk`, `opentelemetry-exporter-jaeger` as optional 
>  1. Install the latest `Twisted == 22.8.0` so it shows up in `poetry.lock`
> 
> ---
> 
> Here are my reproducible steps for trying to get the opentelemetry deps installed:
> 
> ```shell
> # Stop the current virtualenv if active
> $ deactivate
> 
> # Find the venv for poetry
> $ poetry env info --path
> # Remove all of the files from the current environment
> $ rm -rf $(poetry env info --path)
> # Remove the evil egg
> $ rm matrix_synapse.egg-info/
> 
> # Reactivate Poetry shell
> $ poetry shell
> # Install everything
> $ poetry install --extras all
> 
> # Install the new dependencies
> $ poetry add opentelemetry-api opentelemetry-sdk opentelemetry-exporter-jaeger --optional
> ```
> 
> End up with `pyproject.toml` and `poetry.lock` updated but they don't exist in my venv (`/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages`) and can't import from the `python` in the venv,
> 
> ```shell
> $ . /Users/eric/Documents/github/element/synapse/.venv/bin/activate
> 
> $ python
> Python 3.9.13 (main, May 24 2022, 21:28:31)
> [Clang 13.1.6 (clang-1316.0.21.2)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import opentelemetry
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> ModuleNotFoundError: No module named 'opentelemetry'
> ```
> 
>

Reply:
I think there's some confusion here about virtualenvs.

By default, poetry will manage virtualenvs by itself. It stores them in its own cache directory: see `poetry env list --full-path`. If you're not in a virtualenv when you `poetry shell`, this is where you'll be activated into. I think you should be able to see `opentelemetry-api opentelemetry-sdk opentelemetry-exporter-jaeger` in there.

However, it looks like you have a previously-existing venv in your Synapse checkout. If you want to continue using this that's fine: Poetry can manage this virtualenv too. But you have to activate it before running any poetry commands in order to do so.

Does that make sense and/or fix your problem?

(Please LMK if the text of https://matrix-org.github.io/synapse/latest/development/dependencies.html or eleswhere could be improved here.)

Reply:
@DMRobertson The Poetry venv is the one in my Synapse directory which is the one I clean up and re-create over and over:

```
$ poetry env info --path
/Users/eric/Documents/github/element/synapse/.venv
```

Same thing with the command you mention:
```
$ poetry env list --full-path
/Users/eric/Documents/github/element/synapse/.venv (Activated)
```

`poetry shell` creates the venv and "activates" it but doesn't activate in terms of my terminal and interacting as with that Python version like calling `activate` does.

Reply:
Resolved this out of band in #synapse-dev, https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$YBWbXoErCMnRDui-lpsjs-kBpEyaki1US-NpSu2PkyQ?via=matrix.org&via=element.io&via=beeper.com

> @DMRobertson jumped on a call with me and we made some sane looking changes! The linter is still not happy but that might be a separate problem I need to figure out (commented out that part of the CI for now)
> 
> The key things we did different:
> 
> 1. Edit `pyproject.toml` with the OTEL optional deps and add it to the `all` groups, etc before doing any base install (`poetry install --extras all`)
> 2. Updating and verifying no extra changes with `poetry lock --no-update`
> 3. `poetry update twisted` to get the latest Twisted in `poetry.lock`
> 4. Now doing `poetry install --extras all` from a fresh venv does the right thing ✅
>
> Thanks for all the help dmr and Erik !